server/resp/src/test/java/org/infinispan/server/resp/RespBlockingCommandsTest.java

```diff
@@ -0,0 +1,280 @@
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.infinispan.commons.util.concurrent.CompletableFutures;
+import org.infinispan.configuration.cache.CacheMode;
+import org.infinispan.configuration.cache.ConfigurationBuilder;
+import org.infinispan.notifications.Listener;
+import org.infinispan.notifications.cachelistener.CacheNotifier;
+import org.infinispan.notifications.cachelistener.CacheNotifierImpl;
+import org.infinispan.notifications.cachelistener.ListenerHolder;
+import org.infinispan.notifications.cachelistener.annotation.CacheEntryCreated;
+import org.infinispan.notifications.cachelistener.event.CacheEntryEvent;
+import org.infinispan.server.resp.commands.list.blocking.BLPOP;
+import org.infinispan.server.resp.test.RespTestingUtil;
+import org.infinispan.test.TestingUtil;
+import org.mockito.stubbing.Answer;
+import org.testng.annotations.Factory;
+import org.testng.annotations.Test;
+
+import io.lettuce.core.KeyValue;
+import io.lettuce.core.RedisCommandExecutionException;
+import io.lettuce.core.RedisFuture;
+import io.lettuce.core.api.async.RedisAsyncCommands;
+import io.lettuce.core.api.sync.RedisCommands;
+
+/**
+ * Class for blocking commands tests.
+ *
+ * @author Vittorio Rigamonti
+ * @since 15.0
+ */
+@Test(groups = "functional", testName = "server.resp.RespBlockingCommandsTest")
+public class RespBlockingCommandsTest extends SingleNodeRespBaseTest {
+
+   private CacheMode cacheMode = CacheMode.LOCAL;
+   private boolean simpleCache;
+
+   @Factory
+   public Object[] factory() {
+      return new Object[] {
+            new RespBlockingCommandsTest(),
+            new RespBlockingCommandsTest().simpleCache()
+      };
+   }
+
+   RespBlockingCommandsTest simpleCache() {
+      this.cacheMode = CacheMode.LOCAL;
+      this.simpleCache = true;
+      return this;
+   }
+
+   @Override
+   protected String parameters() {
+      return "[simpleCache=" + simpleCache + ", cacheMode=" + cacheMode + "]";
+   }
+
+   @Override
+   protected void amendConfiguration(ConfigurationBuilder configurationBuilder) {
+      if (simpleCache) {
+         configurationBuilder.clustering().cacheMode(CacheMode.LOCAL).simpleCache(true);
+      } else {
+         configurationBuilder.clustering().cacheMode(cacheMode);
+      }
+   }
+
+   private RedisFuture<KeyValue<String, String>> registerBLPOPListener(RedisAsyncCommands<String, String> redis,
+         long timeout, String... keys) {
+      RedisFuture<KeyValue<String, String>> rf = redis.blpop(timeout, keys);
+      CacheNotifierImpl<?, ?> cni = (CacheNotifierImpl<?, ?>) TestingUtil.extractComponent(cache, CacheNotifier.class);
+      // If there's a listener ok otherwise
+      // if rf is done an error during listener registration has happend
+      // no need to wait anymore. test will fail
+      eventually(() -> cni.getListeners().stream()
```

Comment:
While this provides deterministic behavior of registering the listener, we really need additional tests so that we can test what happens when a listener is not yet installed but a poll event finds a value and also what happens if both the poll event and the listener find the same value.

Reply:
not sure I can write tests like these, let me see

Reply:
We also need a test where we have 2 BLPOP listeners waiting on the same key and see what happens if we add one or two elements to that key. For the one case only one listener should complete and the two case they should both have different values.

Reply:
Tested in:
testBlpopTwoListenersWithValues
testBlpopTwoListenersOneTimeout